gitindex/index.go

```
@@ -532,6 +539,229 @@ func newIgnoreMatcher(tree *object.Tree) (*ignore.Matcher, error) {
+
+	existingRepository, err := options.BuildOptions.RepositoryMetadata()
+	if err != nil {
+		return nil, nil, nil, nil, fmt.Errorf("failed to get repository metadata: %w", err)
+	}
+
+	if existingRepository == nil {
+		return nil, nil, nil, nil, fmt.Errorf("no existing shards found for repository")
+	}
+
+	// Check to see if the branch set is consistent with what we last indexed.
+	// If it isn't consistent, that we can't proceed with a delta build (and the caller should fall back to a
+	// normal one).
+
+	indexState := build.CompareBranches(existingRepository.Branches, options.BuildOptions.RepositoryDescription.Branches)
+	if indexState != build.IndexStateBranchVersion {
+		var existingBranchNames []string
+		for _, b := range existingRepository.Branches {
+			existingBranchNames = append(existingBranchNames, b.Name)
+		}
+
+		var optionsBranchNames []string
+		for _, b := range options.BuildOptions.RepositoryDescription.Branches {
+			optionsBranchNames = append(optionsBranchNames, b.Name)
+		}
+
+		existingBranchList := strings.Join(existingBranchNames, ", ")
+		optionsBranchList := strings.Join(optionsBranchNames, ", ")
+
+		return nil, nil, nil, nil, fmt.Errorf("requested branch set in build options (%q) != branch set found on disk (%q) - branch set must be the same for delta shards", optionsBranchList, existingBranchList)
+	}
+
+	// branch => (path, sha1) => repo.
+	repos = map[fileKey]BlobLocation{}
+
+	// fileKey => branches
+	branchMap = map[fileKey][]string{}
+
+	// Branch => Repo => SHA1
+	//branchVersions := map[string]map[string]plumbing.Hash{}
+
+	// branch name -> git worktree at most current commit
+	branchToCurrentTree := make(map[string]*object.Tree, len(options.Branches))
+
+	for _, b := range options.Branches {
+		commit, err := getCommit(repository, options.BranchPrefix, b)
+		if err != nil {
+			return nil, nil, nil, nil, fmt.Errorf("getting last current commit for branch %q: %w", b, err)
+		}
```

the other code path we allow `options.AllowMissingBranch` to be true. Can you fail early if this is true.

---

In this circumstance, what do you mean by failing early? Are you saying that we shouldn't run prepareDeltaBuild _at all_ if options.AllowMissingBranch is set?

---

yes, right now we don't support it.

---

because any error we return means do normal build, we fallback to normal build and respect this option. Therefor we don't need to fail early since we do in fact support this option.